qkit/drivers/ZHInst_Virtual_VNAv2.py

```diff
@@ -0,0 +1,166 @@
+            return np.abs(signal), np.angle(signal)
+        else:
+            return self._i_q_data[:, 0], self._i_q_data[:, 1]
+
+    @QkitFunction
+    def _get_shfsg_frequencies(self) -> tuple((float, float)):
+        """
+        Returns (SHF-Upconversion center frequency, Digital Modulation Frequency)
+        """
+        shfsg_out_freq = self.get_freqpoints()[0] - MINIMUM_UHFQA_FREQUENCY
+        base_mod = min(int(shfsg_out_freq / 0.5e9) * 0.5e9, 8.0e9) # The maximum CF for the SHFSG
+        return base_mod, shfsg_out_freq - base_mod
+
+    @QkitFunction
+    def pre_measurement(self):
+        self.uhfqa.nodetree.qa.result.source(2)
+        self.uhfqa.nodetree.sigouts[0].on(1)
+        self.uhfqa.nodetree.sigouts[1].on(1)
+
+        self.uhfqa.nodetree.awg.outputs[1].amplitude(-1.0)
+        self.uhfqa.arm(length=self.repetitions, averages=1)
+        # Configure output
+        base, digital = self._get_shfsg_frequencies()
+        out_channel = self.shfsg.nodetree.sgchannels[7]
+        synth = self.shfsg.nodetree.synthesizers[out_channel.synthesizer()]
+
+        out_channel.output.range(self.output_power)
+        synth.centerfreq(base)
+        out_channel.awg.modulation.enable(1)
+        out_channel.oscs[0].freq(digital)
+
+        self.uhfqa.compile_program(
+            sequence_type=SequenceType.PULSED_SPEC,
+            period=self.period,
+            repetitions=self.repetitions,
+            trigger_mode=TriggerMode.NONE,
+            alignment=Alignment.START_WITH_TRIGGER
+        )
+        
+        self.shfsg.nodetree.sgchannels[7].output.on(1)
+
+        self.uhfqa.nodetree.qa.result.source(2)
+        self.uhfqa.nodetree.sigouts[0].on(1)# Enable the outputs
+        self.uhfqa.nodetree.sigouts[1].on(1)
+        self.uhfqa.nodetree.awg.outputs[1].amplitude(-1.0)
+
+        self._i_q_data = np.zeros(shape=(len(self.get_freqpoints()), 2), dtype=complex)
+
+    @QkitFunction
+    def start_measurement(self):
+        print("Run")
+        # Configure upper side band modulation
+        out_channel = self.shfsg.nodetree.sgchannels[7]
+        out_channel.sine.oscselect(0)
+        out_channel.sine.i.enable(True)
+        out_channel.sine.i.sin.amplitude(0.0)
+        out_channel.sine.i.cos.amplitude(0.9 * self.amplitude_factor)
+        out_channel.sine.q.enable(True)
+        out_channel.sine.q.sin.amplitude(0.9 * self.amplitude_factor)
+        out_channel.sine.q.cos.amplitude(0.0)
+        self.uhfqa.arm(length=self.repetitions, averages=1)
+
+        base, digital = self._get_shfsg_frequencies()
+        shfsg_frequency = base + digital
+        uhfqa_frequencies = self.get_freqpoints() - shfsg_frequency
+        assert np.all(uhfqa_frequencies < 600e9), "Exceeding maximum UHFQA frequency!"
+
+        for i, f in enumerate(uhfqa_frequencies):
+            self.uhfqa.nodetree.osc.freq(f)  # set modulation frequency
+            for j in (0, 1):
+                self.uhfqa.set_readout_frequency(f, j)
```

Comment:
This call costs time and might be unnecessary. In spectroscopy mode, the signal is (at least according to the documentation) automatically demodulated with the local oscillator.